### PR TITLE
Explore charts

### PR DESCRIPTION
**Pull request by KhadijaMahanga**
## Description
-  Indicator Title
- Chart - Bar, Stacked, Donut, Line
- Chart tooltip
- Chart Embed

Chart:
We are using [Vega Charts](https://vega.github.io/vega/) for our charting system and here is how it [works](https://observablehq.com/@vega/how-vega-works). Sample tutorials are [stacked-chart](https://vega.github.io/vega/examples/stacked-bar-chart/) ,[ bar chart](https://vega.github.io/vega/tutorials/bar-chart/)

Chart Type:
- Default chart is bar chart. To change chart type, update/add the `chart_configuration` on indicator object (see sample on stories). eg: `chart_type: "stacked"` or `chart_type: "donut"`. Currently, stacked chart type has another property that need to be added called `stacked_field`, which points to the field or column that sort the stack.

```
indicator: {
id: 23,
title: "chart A",
description: "",
chart_configuration: {
      chart_type: "stacked",
      stacked_field: "gender",
      ....
},
data: [...]

```

Fixes # (issue)

## Type of change
- [x] New feature (non-breaking change which adds functionality)

### Explore Page
Peek 2021-10-01 09-02

### Bar Charts
Peek 2021-09-29 13-28

### Stacked Charts
Peek 2021-09-29 14-02

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/components/HURUmap/Chart/donutScope.js [hunk 1]
```diff
@@ -0,0 +1,206 @@
+// import { xAxis, xScale } from "./properties";
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/donutScope.js:1]: So many comments

File: src/components/HURUmap/Chart/donutScope.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+// import { createFiltersForGroups } from "./utils";
+
+const PERCENTAGE_TYPE = "percentage";
+const VALUE_TYPE = "value";
+const graphValueTypes = {
+  Percentage: PERCENTAGE_TYPE,
+  Value: VALUE_TYPE,
+};
+
+export default function configureDonutchart(data, metadata, config) {
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/donutScope.js:11]:
```suggestion
export default function configureDonutChart(data, metadata, config) {
```
2 words, right?

File: src/components/HURUmap/Chart/index.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  sourceDiv: {
+    margin: `${typography.pxToRem(20)} 0`,
+  },
+  source: {
+    fontSize: typography.pxToRem(13),
+    lineHeight: 20 / 13,
+    color: "#666666",
+    display: "inline-flex",
+  },
+  link: {
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/index.js:22]:
```suggestion
  source: {
    margin: `${typography.pxToRem(20)} 0`,
  },
  sourceTitle: {
    fontSize: typography.pxToRem(13),
    lineHeight: 20 / 13,
    color: "#666666",
    display: "inline-flex",
  },
  sourceLink: {
```

File: src/components/HURUmap/Chart/index.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    id,
+    description,
+    metadata: { source, url },
+  } = indicator;
+
+  const containerId = `chart-container-${id}`;
+
+  useEffect(() => {
+    async function plotChart(data) {
+      const result = await renderChart(`#${containerId}`, data);
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/index.js:44]: Doesn't support server-side rendering I take?

File: src/components/HURUmap/Chart/renderChart.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import embed from "vega-embed";
+
+import configureDonutchart from "./donutScope";
+import configureBarchart from "./scope";
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/renderChart.js:4]: Why do we call them "scope" and then rename `configureXXXX` on import? Asking for a friend.

File: src/components/HURUmap/ChartTooltip/index.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+          {label}
+        </Typography>
+        {percentValue && (
+          <Typography variant="body2" component="div" className={classes.text}>
+            {percentValue}
+          </Typography>
+        )}
+        <Typography variant="body2" component="div" className={classes.text}>
+          {value}
+        </Typography>
```
Comment (kilemensi) [inline src/components/HURUmap/ChartTooltip/index.js:59]: Should be **SemiBold**

File: src/components/HURUmap/ChartTooltip/index.stories.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+const Template = ({ ...args }) => <ChartTooltip {...args} />;
+
+export const Default = Template.bind({});
+
+Default.args = {
+  label: "15-24",
+  value: "1456000",
+  percentValue: "45%",
```
Comment (kilemensi) [inline src/components/HURUmap/ChartTooltip/index.stories.js:43]:
```suggestion
  title: "15-24",
  value: "1456000",
  formattedValue: "45%",
```
I'm assuming we need to keep naming consistent with the other components.

File: src/components/HURUmap/IndicatorTitle/Download.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  };
+
+  return (
+    <Grid container className={classes.root}>
+      <Grid item container xs={12} className={classes.row}>
+        {["Percentage", "Value"].map((v) => (
+          <Grid item xs={6} index={v} className={classes.button}>
+            <ButtonBase
+              className={clsx(classes.text)}
+              onClick={() => setChartValue(v)}
```
Comment (kilemensi) [inline src/components/HURUmap/IndicatorTitle/Download.js:116]: Doesn't work for me

File: src/components/HURUmap/IndicatorTitle/Download.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+      </Grid>
+      <Grid item xs={12} className={clsx(classes.row, classes.layout)}>
+        <Typography className={classes.text}>Layout option:</Typography>
+      </Grid>
+      <Grid item container className={classes.row}>
+        {["Layout1", "Layout2"].map((p) => (
+          <Grid item xs={6} index={p} className={classes.button}>
+            <ButtonBase
+              className={classes.text}
+              onClick={() => setChartValue(p)}
```
Comment (kilemensi) [inline src/components/HURUmap/IndicatorTitle/Download.js:146]: Doesn't work or I don't know what these Layout options are all about.

File: src/components/HURUmap/IndicatorTitle/Download.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+      </Grid>
+      <Grid item xs={12} className={clsx(classes.row, classes.header)}>
+        <Typography className={classes.text}>Download data as:</Typography>
+      </Grid>
+      <Grid item container className={classes.row}>
+        {["CSV", "XLSX", "JSON"].map((d) => (
+          <Grid item xs={4} index={d} className={classes.button}>
+            <ButtonBase
+              className={classes.text}
+              onClick={(e) => handleDataDownload(e, d)}
```
Comment (kilemensi) [inline src/components/HURUmap/IndicatorTitle/Download.js:161]: JSON doesn't work.

File: src/components/HURUmap/IndicatorTitle/Share.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+function Share({ title, spec, ...props }) {
+  const classes = useStyles(props);
+
+  const code = `<!DOCTYPE html>
+  <html>
+    <head>
+      <title>${title}</title>
+      <script src="https://cdn.jsdelivr.net/npm/vega@5.10.1"></script>
+      <script src="https://cdn.jsdelivr.net/npm/vega-lite@4.10.1"></script>
+      <script src="https://cdn.jsdelivr.net/npm/vega-embed@6.5.2"></script>
```
Comment (kilemensi) [inline src/components/HURUmap/IndicatorTitle/Share.js:62]: Lets pin these to ether minor or major versions only i.e. `@4.10` or `@4`

**Review by kilemensi — COMMENTED** [review]
Massive effort 🙌🏾 

-----------------------

1. Can you add a bit of description of how the charting system works here (scopes, config, render, format, etc.)? Links to resources that the rest of the team can use to get up to speed will also be greatly appreciated.
2. Color/groups/etc., do we know how we're going to be handling this moving foward or not yet?
3. Still not sure how to easily switch chart types in the storybook.

File: src/components/HURUmap/IndicatorTitle/Download.js (hunk 9 above)
Comment (KhadijaMahanga) [reply src/components/HURUmap/IndicatorTitle/Download.js:146]: @kilemensi pierro, just explained it to me yesterday. I will have a story for it

File: src/components/HURUmap/Chart/index.js (hunk 4 above)
Comment (KhadijaMahanga) [reply src/components/HURUmap/Chart/index.js:44]: @kilemensi I think I saw somewhere with an SSR example.

File: src/components/HURUmap/Chart/BarChartScope.js [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+
+  const { primary_group: primaryGroup } = metadata;
+
+  if (xTicks) {
+    xAxis.tickCount = xTicks;
+  }
+
+  const { signals: filterSignals, filters } = createFiltersForGroups(
+    metadata.groups
+  );
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/BarChartScope.js:35]:
```suggestion
  const { primary_group: primaryGroup } = metadata;
  if (xTicks) {
    xAxis.tickCount = xTicks;
  }
  const { signals: filterSignals, filters } = createFiltersForGroups(
    metadata.groups
  );
```

File: src/components/HURUmap/Chart/BarChartScope.js [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+      {
+        name: "yscale",
+        type: "band",
+        domain: { data: "data_formatted", field: { signal: "mainGroup" } },
+        range: { step: { signal: "y_step" } },
+        padding: 0.1,
+      },
+      xScale,
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/BarChartScope.js:171]:
```suggestion
```

File: src/components/HURUmap/Chart/BarChartScope.js [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        scale: "yscale",
+        domainOpacity: 0.5,
+        labelOpacity: 0.5,
+        tickSize: 0,
+        labelPadding: 6,
+        zindex: 1,
+      },
+      xAxis,
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/BarChartScope.js:184]:
```suggestion
```

File: src/components/HURUmap/Chart/DonutChartScope.js [hunk 15]
```diff
@@ -0,0 +1,158 @@
+// import { xAxis, xScale } from "./properties";
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/DonutChartScope.js:1]:
```suggestion
```

File: src/components/HURUmap/Chart/DonutChartScope.js [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        formatting: percentageFormatting,
+        minX: percentageMinX,
+        maxX: percentageMaxX,
+      },
+    },
+  } = config;
+
+  const { primary_group: primaryGroup } = metadata;
+
+  const { filters } = createFiltersForGroups(metadata.groups);
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/DonutChartScope.js:26]:
```suggestion
  const { primary_group: primaryGroup } = metadata;
  const { filters } = createFiltersForGroups(metadata.groups);
```

File: src/components/HURUmap/Chart/DonutChartScope.js [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+  return {
+    $schema: "https://vega.github.io/schema/vega/v5.json",
+    description: "A basic donut chart example.",
+    width: 200,
+    height: 200,
+    autosize: "none",
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/DonutChartScope.js:34]:
```suggestion
```

File: src/components/HURUmap/Chart/DonutChartScope.js [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+      {
+        name: "valueMaxX",
+        value: valueMaxX !== "default" ? valueMaxX : undefined,
+      },
+      {
+        name: "valueMinX",
+        value: valueMinX !== "default" ? valueMinX : undefined,
+      },
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/DonutChartScope.js:74]:
```suggestion
```

File: src/components/HURUmap/Chart/DonutChartScope.js [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+        title: "Legends",
+        orient: "none",
+        symbolType: "circle",
+        padding: { value: 10 },
+        encode: {
+          symbols: { enter: { fillOpacity: { value: 1 } } },
+        },
+      },
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/DonutChartScope.js:128]:
```suggestion
```

File: src/components/HURUmap/Chart/DonutChartScope.js [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+    scales: [
+      {
+        name: "color",
+        type: "ordinal",
+        domain: { data: "table", field: { signal: "mainGroup" } },
+        range: { scheme: "category20" },
+      },
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/DonutChartScope.js:137]:
```suggestion
```

File: src/components/HURUmap/Chart/StackedChartScope.js [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+
+  const { primary_group: primaryGroup } = metadata;
+
+  if (xTicks) {
+    xAxis.tickCount = xTicks;
+  }
+
+  const { signals: filterSignals, filters } = createFiltersForGroups(
+    metadata.groups
+  );
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/StackedChartScope.js:36]:
```suggestion
  const { primary_group: primaryGroup } = metadata;
  if (xTicks) {
    xAxis.tickCount = xTicks;
  }
  const { signals: filterSignals, filters } = createFiltersForGroups(
    metadata.groups
  );
```

File: src/components/HURUmap/Chart/StackedChartScope.js [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+        name: "color",
+        type: "ordinal",
+        range: "category",
+        domain: {
+          data: "data_formatted",
+          field: stackedField,
+        },
+      },
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/StackedChartScope.js:200]:
```suggestion
```

File: src/components/HURUmap/Chart/StackedChartScope.js [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+          },
+          symbols: {
+            update: {
+              stroke: { value: "transparent" },
+            },
+          },
+        },
+      },
+    ],
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/StackedChartScope.js:236]:
```suggestion
```

File: src/components/HURUmap/Chart/configureScope.js [hunk 24]
```diff
@@ -0,0 +1,31 @@
+import BarChartScope from "./BarChartScope";
+import DonutChartScope from "./DonutChartScope";
+import StackedChartScope from "./StackedChartScope";
+
+export default function configureScope(indicator) {
+  const configuration = indicator?.chart_configuration;
+
+  let vegaSpec;
+
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/configureScope.js:9]:
```suggestion
```

File: src/components/HURUmap/Chart/utils.js [hunk 25]
```diff
@@ -0,0 +1,31 @@
+export const slugify = (string) => {
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/utils.js:1]: [PT](https://github.com/CodeForAfrica/PromiseTracker/blob/e70c0928046110fd5104be807d3bd410ba59d93d/src/utils.js#L1-L18) `slugify` seem to handle even special characters.

File: src/components/HURUmap/IndicatorTitle/Download.js (hunk 8 above)
Comment (kilemensi) [reply src/components/HURUmap/IndicatorTitle/Download.js:116]: Still doesn't work for me

File: src/theme/index.js [hunk 26]
```diff
@@ -50,6 +52,10 @@ const theme = createTheme({
       secondary: "#FFFFFF",
       hint: "#1C2031",
     },
+    chart: {
+      text: "#707070",
+      category: CHART_PRIMARY_COLOR_SCHEME,
+    },
```
Comment (kilemensi) [inline src/theme/index.js:58]: Why not replicate MUI approach i.e. `chart: { primary: ..., secondary: ..., text: { primary: ..., secondary: ..} }`?

File: src/components/HURUmap/IndicatorTitle/index.js [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+    },
+  ];
+
+  return (
+    <div className={classes.root}>
+      <Grid container justifyContent="space-between">
+        <Grid item xs={12} md={8}>
+          <Typography variant="h6">{title}</Typography>
+        </Grid>
+        <Grid item xs={12} md={4} container justifyContent="flex-end">
```
Comment (kilemensi) [inline src/components/HURUmap/IndicatorTitle/index.js:73]: In mobile, actions should be `flex-start`
Screenshot from 2021-09-24 13-34-46

File: src/components/HURUmap/IndicatorTitle/index.js [hunk 28]
```diff
@@ -0,0 +1,95 @@
...
+    },
+    {
+      id: "act-download",
+      header: "Chart value as:",
+      children: <Download title={title} {...props} />,
+      icon: <DownloadIcon />,
+    },
+    {
+      id: "act-share",
+      header: "Share chart via:",
```
Comment (kilemensi) [inline src/components/HURUmap/IndicatorTitle/index.js:61]: Shouldn't these come from CMS/Backend/config/etc.

**Review by kilemensi — COMMENTED** [review]
👍🏾 

---

1. Total chart width (including padding, etc.): desktop 768, tablet 600 and mobile 350,
2. Support for dotted line (national average),
3. (Stacked) Bar chart should be up-right in mobile,
4. Why are we still not using the custom Tooltip component,
5. Changing config to `donut` for chart type doesn't work in storybook.
6. Some of these config/scopes look very similar especially for bar and stacked bar... may be we can have a base config and use deepmerge to override when required?

File: src/components/HURUmap/IndicatorTitle/Download.js (hunk 9 above)
Comment (KhadijaMahanga) [reply src/components/HURUmap/IndicatorTitle/Download.js:146]: Added a story on [PT]([url](https://www.pivotaltracker.com/story/show/179750888)) for this

**Review by kilemensi — COMMENTED** [review]
👍🏾 @KhadijaMahanga ... looks dope.

--------------------

Quick review:

1. I can't click on any of the action buttons. download, description, etc.
2. Horizontally, our charts need to be a bit _wider_ to match the designs.
S1V
3. Vertically, I think we've nailed it. My only question is do the charts use the same font as the rest of the site?
S1H
4. After every chart, the subcategory top border disappears.
Screenshot from 2021-10-01 10-45-33

**Comment by KhadijaMahanga** [comment]
@kilemensi 
1. Any idea what could be swallowing those button events? I can't seem to find the source
2.
3. Charts are using Poppins regular too
4. Because we don't have top border on subcategories. I will add one

**Comment by kilemensi** [comment]
>     1. Any idea what could be swallowing those button events? I can't seem to find the source
> 
I think events are fine @KhadijaMahanga but `z-index` is not set properly ... since panel is at [drawer](https://github.com/CodeForAfrica/PesaYetu/blob/bbe8a46a0d6b5a6d5e37493386e136749958e8eb/src/components/HURUmap/Panel/useStyles.js#L66) `z-index`, then the pop-ups must also be at the same `z-index`. Otherwise, they'll appear below the panel.

>     3. Charts are using Poppins regular too
> 
Nice nice.

**Review by KobbyMmo — APPROVED** [review]
Looks So Beautiful @KhadijaMahanga 
🎸 🚀 👯 💃 🕺 

A few comments 
1. the layout was not changing when I clicked the changing layout. But I think the XD is like that as well. 
2. Changing percentages and values didn't affect the chart.

But overall, It looks really Goooooooooooood!!!!

**Review by nerdlet — APPROVED** [review]
Looks really good 🚀 🚀  @KhadijaMahanga

File: src/components/HURUmap/Chart/VerticalStackedChartScope.js [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+        domain: { data: "data_formatted", field: { signal: "mainGroup" } },
+        range: { step: { signal: "x_step" } },
+        padding: 0.15,
+      },
+      {
+        name: "yscale",
+        type: "linear",
+        domain: { data: "data_formatted", field: "y1" },
+        // domainMin: { signal: "domainMin" },
+        // domainMax: { signal: "domainMax" },
```
Comment (kilemensi) [inline src/components/HURUmap/Chart/VerticalStackedChartScope.js:175]: ?

**Review by kilemensi — APPROVED** [review]
🚀 🚀

**Comment by KhadijaMahanga** [comment]
> Looks So Beautiful @KhadijaMahanga guitar rocket dancers woman_dancing man_dancing
> 
> A few comments
> 
> 1. the layout was not changing when I clicked the changing layout. But I think the XD is like that as well.
> 2. Changing percentages and values didn't affect the chart.
> 
> But overall, It looks really Goooooooooooood!!!!

@KobbyMmo I have separate stories for the above

**Pull request merged**